miniapps/autodiff/admfem.hpp

```diff
@@ -0,0 +1,587 @@
+                rr[jj].setGradient(0.0);
+            }
+            tape.reset(pos);
+        }
+#endif
+    }
+};
+
+template<template<typename, typename, typename, int, int> class CTD
+         , int state_size=1, int param_size=0>
+class QFunctionAutoDiff
+{
+public:
+
+    /// Evaluates a function for arguments vparam and uu.
+    /// The evaluatin is based on the operator() in the
+    /// user provided functor CTD.
+    double QEval(const Vector &vparam, Vector &uu)
+    {
+        CTD<double, const Vector, Vector, state_size, param_size> tf;
+        return tf(vparam,uu);
+    }
+
+    /// Provides the same functionality as QGrad.
+    void QVectorFunc(const Vector &vparam, Vector &uu, Vector &rr)
+    {
+        QGrad(vparam,uu,rr);
+    }
+
+    /// Returns the first derivative of CTD(...) with
+    /// respect to the active arguments proved in vector uu.
+    /// The length of rr is the same as for uu.
+    void QGrad(const Vector &vparam, Vector &uu, Vector &rr)
+    {
+
+#ifdef MFEM_USE_ADFORWARD
+        // use forward mode
+        CTD<ad::ADFloatType, const Vector, ad::ADVectorType,
+                        state_size, param_size> tf;
+        rr.SetSize(state_size);
+        {
+            ad::ADVectorType aduu(state_size);
+            for(int i=0;i<state_size;i++){
+                aduu[i].setValue(uu[i]);
+                aduu[i].setGradient(0.0);
+            }
+
+            ad::ADFloatType rez;
+
+            for(int ii=0;ii<state_size;ii++){
+                aduu[ii].setGradient(1.0);
+                rez=tf(vparam,aduu);
+                rr[ii]=rez.getGradient();
+                aduu[ii].setGradient(0.0);
+            }
+        }
+
+#else
+        typedef codi::RealReverse ADFType;
+        typedef TADVector<ADFType> ADFVector;
+
+        CTD<ADFType, const Vector, ADFVector, state_size, param_size> tf;
+        {
+            ADFVector aduu(state_size);
+            ADFType rez;
+            for(int i=0;i<state_size;i++){
+                aduu[i]=uu[i];
+            }
+
+            ADFType::TapeType& tape =ADFType::getGlobalTape();
+            typename ADFType::TapeType::Position pos=tape.getPosition();
+
+            tape.setActive();
+            for(int ii=0;ii<state_size;ii++){ tape.registerInput(aduu[ii]); }
+
+            rez=tf(vparam,aduu);
+            tape.registerOutput(rez);
+            tape.setPassive();
+
+            rez.setGradient(1.0);
+            tape.evaluate();
+            for(int i=0;i<state_size;i++){
+                rr[i]=aduu[i].getGradient();
+            }
+            tape.reset(pos);
+        }
+#endif
+    }
+
+    /// Provides same functionality as QHessian.
+    void QJacobian(mfem::Vector &vparam, mfem::Vector &uu, mfem::DenseMatrix &jac)
+    {
+        QHessian(vparam,uu,jac);
+    }
+
+    /// Returns the Hessian of CTD(...) in the dense matrix hh.
+    /// The dimensions of jac are state_size x state_size, where state_size is the
+    /// length of vector uu.
```

Comment:
The comments say that the Hessian will be returned in the dense matrix `hh`, but there is no argument named `hh`. The method's implementation makes no references to `hh`.
Are there any use cases where storing the Hessian in a dense matrix may be impossible due to the large number of variables?  Are there any plans for sparse Jacobian/Hessians?

Reply:
> Are there any plans for sparse Jacobian/Hessians?

If I'm understanding @bslazarov's approach correctly, the idea is to apply AD to functions evaluated at quadrature points (what [libCEED calls "Q functions"](https://libceed.readthedocs.io/en/latest/libCEEDapi/#theoretical-framework) under the assumption that the number of differentiable parameters at each quadrature point is small. This approach is elegant because it enables one to leverage the sparsity structure of the assembled finite element equations to construct an assembled Jacobian with respect to the differentiable parameters -- without having to manage explicitly the sparsity structure of that assembled Jacobian. Consequently, I don't see why there would be any plans for sparse Jacobians/Hessians as distinct entities for AD in MFEM.

> Are there any use cases where storing the Hessian in a dense matrix may be impossible due to the large number of variables?

There are use cases where storing the Hessian would be impossible, but most of those use cases would be intractable regardless of approach because the assembled Jacobian with respect to the differentiable parameters would be dense, so I don't think those use cases are of concern.

Reply:
Documentation is modified. 

The AD classes should be used at integration point level for evaluating energy, residuals or Hessian matrices. Thus the matrices will be quite small (3x3,4x4 may be 12x12 for mixed elasticity) where dense matrices will cover most of the cases. The implementation is not intended to be used for differentiating global objectives and functions.